### DISCUSSION TITLE
Unnecessary system calls

### DISCUSSION THREAD

**Discussion by hamdav** [bevyengine/bevy#3550] [Q&A] [answered]
Hello!

I have just recently taken an interest in Bevy and am trying to learn the basics and philosophy of Bevy and ECS. One thing that is bugging me is that there seems to be a lot of unnecessary system calls. A lot of times in development there are things that can only happen after certain other events. For example, in the [events.rs](https://github.com/bevyengine/bevy/blob/latest/examples/ecs/event.rs) file, the listener function gets called constantly, but it only really has an effect when the event trigger function triggers the event. Would it not be better to call this function directly from the trigger function? Or is the slowdown not perceptible enough that you should prefer the readability increase by generating events?

**Comment by alice-i-cecile** [comment, marked as answer]
So, this is a question of "push" vs "pull" styles for event management. Generally, the overhead of systems is very low, if they're not doing any work, they can be largely ignored.

That said, we're at the brainstorming phase for other ways we can reduce these empty system calls: https://github.com/bevyengine/rfcs/pull/32 is particularly relevant to your question :)

**Reply by cart** [reply]
Yeah the hard problem to solve here is that if you don't want to poll for new events in a system, you need to immediately react to them _when they occur_. However if you want arbitrary reactions to events with arbitrary Bevy ECS data access, you can't react immediately because the ECS data needed for the reaction might already be in use elsewhere.

A simple solution to this problem is to make all event production and consumption single-threaded (which is what single threaded ECS-es like Entitas do). But that would mean throwing away all of the cool multithreading stuff that Bevy ECS gives you. We might add this style of reaction for those who want it, but I'm not (yet) sure it should be a recommended pattern for events generally. We have been discussing middle ground solutions, but this is definitely a "you cant have your cake and eat it too" situation.

**Reply by hamdav** [reply]
Okay, thanks for the quick response :) I can see the benefits of this approach much better now, I do like my multithreading
